core/src/main/java/org/ehcache/core/internal/service/ServiceLocator.java

```diff
@@ -324,7 +325,16 @@ public ServiceLocator build() {
 
     ServiceMap lookupDependenciesOf(ServiceMap resolved, Class<? extends Service> requested) throws DependencyException {
       for (Class<? extends Service> dependency : identifyImmediateDependenciesOf(requested)) {
-        resolved = lookupService(resolved, dependency);
+        try {
```

Comment:
I'm fine with the general concept. The thing I don't get is why blending normal dependencies and optional ones. They could be resolved separately: `identifyImmediateDependenciesOf` and `identifyImmediateOptionalDependenciesOf`. With the exception check only on the later. This is not directly the code of this PR. It is a quirk in the original code. The PR just makes it more obvious.

Reply:
I've refactored the code to introduce `identifyImmediateOptionalDependenciesOf` but I do find the code less readable.

But I'll leave it up to you which form you prefer.